docs/user-guide/app_dependencies.md

```diff
@@ -0,0 +1,121 @@
+# Application dependencies
+
+Starting from Argo CD v2.9, you can natively model dependencies between Applications. This new mechanism is intended to replace the prior way of modeling dependencies when using the app-of-apps pattern with sync waves on the Application custom resources.
+
+Using the new dependency mechanism, users are not bound to the app-of-apps pattern anymore and can define dependencies between applications using application selectors to define a directed, acyclic graph for dependency modeling. This means that an Application which defines one or more dependencies do not only depend on its direct dependencies, but to the dependencies of its dependencies as well.
+
+Application dependencies support both, manual and automatic sync policies.
+
+## Feature state and caveats
+
+This feature is in [beta state](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md#beta). It will be polished across the upcoming releases.
+
+Bug reports and enhancement requests are most welcome.
+
+## Limitations and known issues
+
+A list of things that are good to know before getting started with Application dependencies:
+
+* An Application can only depend on other Applications that use the same AppProject and are in the same namespace as the dependent Application.
```

Comment:
Are there any plans to remove the same namespace restriction in the future? I can understand that the dependent applications should be under the same AppProject, but regarding the namespaces they may be spread over the namespaces, especially in the cluster administrator point of view.

Reply:
nvm, I completely misunderstood the Application namespace and the target namespace that is defined in Application. There may be some use cases because we already have the concept of [Applications in any namespace](https://argo-cd.readthedocs.io/en/stable/operator-manual/app-any-namespace/), but I don't have a requirement about it.

Reply:
I second the part about depending on an `Application` belonging to another `AppProject`. As a specific example we have a bunch of "baseapps" which are managed by the DevOps team, which belongs to a baseapps project. These applications, from a developer's perspective are read-only (they can view these apps, but not do anything apart from that). 

It would be useful to have a dependency between an `Application` belonging to the baseapp project, and an `Application` which belongs to another `AppProject` (in our case, every team has their own dedicated `AppProject`).

A contrived example here (this specific example would be better handled by retrying the dependent application, but just to illustrate the dependency):

1. `aws-load-balancer-controller` belongs to the baseapp project
2. `foo-app` belonging to AppProject A needs `aws-load-balancer-controller` to be up before starting
3. `bar-app` belonging to AppProject B  also needs `aws-load-balancer-controller` to be up before starting

Reply:
I feel this is a valid requirement, and I had initially implemented it, but we need to be careful here. Because in a multi-tenant system, you may not want someone with access in project A to even know about project B, let alone the Applications in that project or their sync/health status.

@crenshaw-dev and I also had a brief discussion about this today, but without a final conclusion. 

How about having that as a setting in argocd-cm (or as a command line switch to the application controller, or somewhere else) that allows cross-project and cross-namespace dependencies and that must be explicitly turned on by the Argo CD administrator?

Reply:
I presume that there's a default AppProject? (It isn't clear from this text, and I'd hate to have to, as an end user, figure out if I have to opt into another feature first before I could use this one.)

Reply:
@jsoref  AppProjects are one of the core concepts of Argo CD, so I feel users should know about it already

Reply:
I have no idea. And I've been using it for years. (Seriously)

Reply:
> How about having that as a setting in argocd-cm (or as a command line switch to the application controller, or somewhere else) that allows cross-project and cross-namespace dependencies and that must be explicitly turned on by the Argo CD administrator?

It makes sense for this to be opt-in by the Argo CD admin. Ideally this would be a configuration in the AppProject itself, something like

```yaml
# This would allow appprojects to depend on the app project declaring the stanza below
permitDependsOnFor:
- appProjectA
- appProjectB
```

I'm not sure how this would interplay with self-serve app projects, so if there's something preventing this from being possible then I guess we would need to configure this directly in the application controller. Or perhaps another CRD which can be referenced by AppProjects which in turn would only be allowed to be managed by Argo CD admins?

Reply:
Sorry for the late reply here, @blakepettersson 

>I'm not sure how this would interplay with self-serve app projects

I don't think self-serve app projects are a good idea, and I think we don't have to consider this scenario. So your example makes sense. I think we might want to bring this enhancement in the next iteration of this feature, if it's something our users want.